backend/stakepoold/rpc/rpcserver/client.go

```diff
@@ -0,0 +1,168 @@
+	client *rpcclient.Client
+	mux    sync.RWMutex
+
+	cfg          *rpcclient.ConnConfig
+	ntfnHandlers *rpcclient.NotificationHandlers
+	wg           sync.WaitGroup
+	stop         chan struct{}
+
+	connected    chan struct{}
+	connectedMux sync.Mutex
+}
+
+// Connected returns a receiving copy of the current connected channel. If
+// disconnected and the channel is not yet blocking, creates a new channel that
+// will be closed on a successful reconnect.
+func (c *Client) Connected() <-chan struct{} {
+	c.connectedMux.Lock()
+	defer c.connectedMux.Unlock()
+	if c.RPCClient().Disconnected() {
+		// Start blocking on connected chan if not already.
+		select {
+		case <-c.connected:
+			c.connected = make(chan struct{})
+		default:
+		}
+	}
+	return c.connected
+}
+
+// RPCClient allows access to the underlying rpcclient by providing a copy of
+// its address.
+func (c *Client) RPCClient() *rpcclient.Client {
+	c.mux.RLock()
+	defer c.mux.RUnlock()
+	return c.client
+}
+
+// New creates a new Client and starts the automatic reconnection handler.
+// Returns an error if unable to construct a new rpcclient.
+func New(cfg *rpcclient.ConnConfig, ntfnHandlers *rpcclient.NotificationHandlers) (*Client, error) {
+	client, err := rpcclient.New(cfg, ntfnHandlers)
+	if err != nil {
+		return nil, err
+	}
+	c := &Client{
+		client:       client,
+		cfg:          cfg,
+		ntfnHandlers: ntfnHandlers,
+		stop:         make(chan struct{}),
+		connected:    make(chan struct{}),
+	}
+	// A closed connected channel indcates successfully connected.
+	close(c.connected)
+	go func(c *Client) {
+		c.wg.Add(1)
+		c.autoReconnect()
+	}(c)
+	return c, nil
+}
+
+// Stop stops automatic reconnections.
+func (c *Client) Stop() {
+	c.stop <- struct{}{}
+	c.wg.Wait()
+	// Stop blocking on connected if blocking, as we will never reconnect again.
+	select {
+	case <-c.connected:
+	default:
+		close(c.connected)
+	}
+}
+
+// autoReconnect waits for a disconnect or stop. On disconnect it attempts to
+// reconnect to the client every connectionRetryInterval.
+//
+// This function must be run as a goroutine.
+func (c *Client) autoReconnect() {
+out:
+	for {
+		select {
+		case <-c.stop:
+			break out
+		case <-time.After(disconnectCheckInterval):
+			if !c.RPCClient().Disconnected() {
+				continue out
+			}
+			// Client is disconnected. Try to reconnect.
+			// Start blocking on connected chan if not already.
+			c.connectedMux.Lock()
+			select {
+			case <-c.connected:
+				c.connected = make(chan struct{})
+			default:
+			}
+			c.connectedMux.Unlock()
+		}
+
+	reconnect:
+		for {
+			select {
+			case <-c.stop:
+				break out
+			default:
+			}
+
+			// Start a new client.
+			client, err := rpcclient.New(c.cfg, c.ntfnHandlers)
+
+			if err != nil {
+				log.Warnf("Failed to connect to %s: %v",
+					c.cfg.Host, err)
+
+				log.Infof("Retrying connection to %s in "+
+					"%s", c.cfg.Host, connectionRetryInterval)
+				time.Sleep(connectionRetryInterval)
```

Comment:
An exponential backoff here would be better to avoid using up file descriptors trying to connect at a constant rate.

Reply:
The interval is consistent with what is being done in dcrwallet:
https://github.com/decred/dcrwallet/blob/51900c2bd053bcca091d030611b6ada482719220/dcrwallet.go#L447-L454

I was going for simple as far as the timing is concerned.  Or perhaps I don't understand the problem.  I will google a bit more.

Reply:
Stopped at 2019-09-13 10:39:29.601 because I need to do something else
no signs of what happened in that issue. Readings from top, df, and lsof all seem normal on both stakepoold and dcrstakepool.

Reply:
It's confusing that "file descriptors" also refer to handles on other os resources like network connections, but you'd get a count for a certain process by looking in `/proc/{pid}/fd`.

I can only hope that what we saw years ago was the result of a bug causing unused connections to hang around eating up such handles.  You might consider lowering the retry time to 250ms and monitoring the processes fd proc folder to see if they accumulate and are not released as expected.

Reply:
good idea, I'll do that before going to sleep and see what happens.

google told me that lsof -p PID would show me a list of "file descriptors"

Reply:
Ah, yes lsof does, sorry.

Reply:
Regardless of whether this file descriptor issue is still an issue, a reconnect interval that increases like 1, 2, 4, 8, 16, 32, 64, 64, 64, 64, ... is pretty simple and not uncommon reconnect behavior.  That said, I'm sure it's fine just to try every 5 seconds too.

Reply:
One hour at 250ms and there was no change in the number of file descriptors.

dcrd's rpcclient does have an increasing timer. I promise to make sure the permanent answer to stakepoold's connection problems includes it.